### ISSUE TITLE
Graph coloring issue?

### ISSUE THREAD

**Issue by johnvtan** [mfekadu/rc#42]
The program `if-test2.rc` contains an example where two variables that are alive at the same time get assigned the same register.

**Comment by mfekadu** [comment]
on the branch **`finishing-bools`**
```racket
> ./run programs/if-should-exit-42.rc
44
```

on the branch **`coloring-fixes`**
```racket
> ./run programs/if-should-exit-42.rc
42
```

```racket
> ./run programs/if-test2.rc
92
```

```racket
> cat programs/if-test2.rc
(program ()
    (let ([x #t])
        (if (not x)
          (+ 3 (if (>= 4 1) 2 0))
          (- 46 (if (eq? x #t) 4 2)))))
```

```assembly
> cat programs/if-test2.s
      .global _main
_main:
      movq %rsp, %rbp
      jmp start
block1347:
      movq %rbx, %rax
      jmp conclusion
block1351:
      movq %rbx, %rbx
      negq %rbx
      movq $46, %rbx    ; this is bad... why do we overwrite rbx?? when the "x" is still alive??
      addq %rbx, %rbx
      jmp block1347
block1352:
      movq $4, %rbx
      jmp block1351
block1353:
      movq $2, %rbx
      jmp block1351
block1348:
      movq $3, %rbx
      addq %rbx, %rbx
      jmp block1347
block1349:
      movq $2, %rbx
      jmp block1348
block1350:
      movq $0, %rbx
      jmp block1348
block1354:
      movq $4, %rax
      cmpq $1, %rax
      setl %al
      movzbq %al, %rbx
      cmpq $0, %rbx
      sete %al
      movzbq %al, %rbx
      movq $1, %rax
      cmpq %rbx, %rax
      je block1349
      jmp block1350
block1355:
      movq $1, %rax
      cmpq %rbx, %rax
      sete %al
      movzbq %al, %rbx
      movq $1, %rax
      cmpq %rbx, %rax
      je block1352
      jmp block1353
start:
      movq $1, %rbx
      cmpq $0, %rbx
      sete %al
      movzbq %al, %rcx
      movq $1, %rax
      cmpq %rcx, %rax
      je block1354
      jmp block1355

conclusion:
      retq
>
```

**Comment by mfekadu** [comment]
```racket
> cat compiler.rkt
...
...
(define (compile prog)
                  (rco-prog
                    (uniquify
                      (shrink
                        (typecheck prog)))))
....
....
> ./compiler.rkt programs/if-test2.rc
(program ()
         (let ((x1338 #t))
           (let ((complex-if1346
                  (let ((complex-op1339 (not x1338)))
                    (if complex-op1339
                        (let ((complex-if1342
                               (let ((complex-op1341 (< 4 1)))
                                 (let ((complex-op1340 (not complex-op1341)))
                                   (if complex-op1340 2 0)))))
                          (+ 3 complex-if1342))
                        (let ((complex-if1345 (let ((complex-op1344 (eq? x1338 #t))) (if complex-op1344 4 2))))
                          (let ((complex-op1343 (- complex-if1345)))
                            (+ 46 complex-op1343)))))))
             complex-if1346)))
```

**Comment by johnvtan** [comment]
```racket
((block
  (#<set: complex-if3053> #<set: complex-if3053> #<set: complex-if30483048> #<set:> #<set:>)
  (label block3054)
  (movq (var complex-if3053) (var complex-if30483048))
  (movq (var complex-if30483048) (reg rax))
  (jmp conclusion)) 

(block
 (#<set: complex-if3052>
         #<set: complex-if3052>
         #<set: complex-if30473046>
         #<set: complex-op30453047>
         #<set: complex-op30453047>
         #<set: complex-op30453047 complex-op3051>
         #<set: complex-op3051>
         #<set: complex-if3053>
         #<set: complex-if3053>)
 (label block3058)
 (movq (var complex-if3052) (var complex-if30473046))
 (movq (var complex-if30473046) (var complex-op30453047))
 (negq (var complex-op30453047))
 (movq (int 46) (var complex-op3051))
(addq (var complex-op30453047) (var complex-op3051)) ; this one is not cool because they end up both in rbx in the asm output
 (movq (var complex-op3051) (var complex-if3053))
 (jmp block3054)) 

(block
 (#<set:> #<set:> #<set: complex-if3052> #<set: complex-if3052>)
 (label block3060)
 (movq (int 2) (var complex-if3052))
 (jmp block3058)) 

(block
 (#<set: complex-if3050> #<set: complex-if3050> #<set: complex-if30443044> #<set: complex-if30443044 complex-op3049> #<set: complex-op3049> #<set: complex-if3053> #<set: complex-if3053>)
 (label block3055)
 (movq (var complex-if3050) (var complex-if30443044))
 (movq (int 3) (var complex-op3049))
 (addq (var complex-if30443044) (var complex-op3049))
 (movq (var complex-op3049) (var complex-if3053))
 (jmp block3054))

(block
 (#<set:> #<set:> #<set: complex-if3050> #<set: complex-if3050>)
 (label block3056)
 (movq (int 2) (var complex-if3050))
 (jmp block3055))

(block
 (#<set:> #<set:> #<set: complex-if3050> #<set: complex-if3050>)
 (label block3057)
 (movq (int 0) (var complex-if3050))
 (jmp block3055))

(block
 (#<set:> #<set:> #<set:> #<set:> #<set:> #<set:> #<set:> #<set: complex-op30423043> #<set:> #<set:> #<set:>)
 (label block3061)
 (cmpq (int 1) (int 4))
 (set l (byte-reg al))
 (movzbq (byte-reg al) (var complex-op30433042))
 (cmpq (int 0) (var complex-op30433042))
 (set e (byte-reg al))
 (movzbq (byte-reg al) (var complex-op30423043))
 (cmpq (var complex-op30423043) (int 1))
 (jmp-if e block3056)
 (jmp block3057))

(block
 (#<set:> #<set:> #<set: complex-if3052> #<set: complex-if3052>)
 (label block3059)
 (movq (int 4) (var complex-if3052))
 (jmp block3058))

(block
 (#<set: x30403040> #<set: x30403040> #<set:> #<set:> #<set: complex-op30463045> #<set:> #<set:> #<set:>)
 (label block3062)
 (cmpq (var x30403040) (int 1))
 (set e (byte-reg al))
 (movzbq (byte-reg al) (var complex-op30463045))
 (cmpq (var complex-op30463045) (int 1))
 (jmp-if e block3059)
 (jmp block3060))

(block
 (#<set:> #<set:> #<set:> #<set: x30403040> #<set: x30403040> #<set: x30403040 complex-op30413041> #<set: x30403040> #<set: x30403040> #<set: x30403040>)
 (label start)
 (movq (int 1) (var x30403040))
 (cmpq (int 0) (var x30403040))
 (set e (byte-reg al))
 (movzbq (byte-reg al) (var complex-op30413041))
 (cmpq (var complex-op30413041) (int 1))
 (jmp-if e block3061)
 (jmp block3062)))
```

**Comment by mfekadu** [comment]
**Here is just that one bad block**
```racket
(define bs `((block
 (,(set 'complex-if3052)
         ,(set 'complex-if3052)
         ,(set 'complex-if30473046)
         ,(set 'complex-op30453047)
         ,(set 'complex-op30453047)
         ,(set 'complex-op30453047 'complex-op3051)
         ,(set 'complex-op3051)
         ,(set 'complex-if3053)
         ,(set 'complex-if3053>))
 (label block3058)
 (movq (var complex-if3052) (var complex-if30473046))
 (movq (var complex-if30473046) (var complex-op30453047))
 (negq (var complex-op30453047))
 (movq (int 46) (var complex-op3051))
(addq (var complex-op30453047) (var complex-op3051)) ; this one is not cool because they end up both in rbx in the asm output
 (movq (var complex-op3051) (var complex-if3053))
 (jmp block3054))))
```

**locals**
```racket
(define locals (set->list (list->set (list 'complex-if3052
'complex-if3052
'complex-if30473046
'complex-op30453047
'complex-op30453047
'complex-op30453047 'complex-op3051
'complex-op3051
'complex-if3053
'complex-if3053))))
```

**program**
```racket
(define p `(program ,locals ,bs))
```


```
> (define useful-live-list (drop (second (first bs)) 1))
> (displayln useful-live-list)
(#<set: complex-if3052> #<set: complex-if30473046> #<set: complex-op30453047> #<set: complex-op30453047> #<set: complex-op3051 complex-op30453047> #<set: complex-op3051> #<set: complex-if3053> #<set: complex-if3053>>)
```

```
> (define instrs (drop (first bs) 2))
> (displayln instrs)
((label block3058) (movq (var complex-if3052) (var complex-if30473046)) (movq (var complex-if30473046) (var complex-op30453047)) (negq (var complex-op30453047)) (movq (int 46) (var complex-op3051)) (addq (var complex-op30453047) (var complex-op3051)) (movq (var complex-op3051) (var complex-if3053)) (jmp block3054))
```

```
(check-equal? (length instrs) (length useful-live-list))
```

```
> (displayln (interference-from-live useful-live-list instrs '())) ; empty graph init
((complex-if3053 #<set:> #<set:>) ; vertex, saturation, edges
 (complex-op3051 #<set:> #<set:>)
 (complex-op30453047 #<set:> #<set: complex-op3051>)
 (complex-op3051 #<set:> #<set: complex-op30453047>)
 (complex-op30453047 #<set:> #<set:>)
 (complex-if30473046 #<set:> #<set:>))
```

```
(define g `((complex-if3053 ,(set) ,(set)) ; vertex, saturation, edges
 (complex-op3051 ,(set) ,(set))
 (complex-op30453047 ,(set) ,(set 'complex-op3051))
 (complex-op3051 ,(set) ,(set 'complex-op30453047))
 (complex-op30453047 ,(set) ,(set))
 (complex-if30473046 ,(set) ,(set))))
```

```
> (displayln locals)
(complex-if30473046 complex-op30453047 complex-if3052 complex-if3053 complex-op3051)
```

## OH NO!!

```
(color-graph g locals)
. . graph-get-saturation: ERROR tried to get edges from vertex { 'complex-if3052 } not in graph
```


**So it seems like this is a new problem where `color-graph` tries to loop over locals but locals is global to the whole program and this particular local vertex `complex-if3052` is not in this block**

**JK... we didn't make the right graph during testing... let's try to fix that**


## Oh okay cool... hmmmm... why all color 0 except for `complex-op3051`
```
> (define locals '(complex-if30473046 complex-op30453047 complex-if3053 complex-op3051))
> (color-graph g locals)
(make-immutable-hash (list (cons 'complex-op3051 1) (cons 'complex-if3053 0) (cons 'complex-op30453047 0) (cons 'complex-if30473046 0)))
```

## OKay... let's keep exploring... maybe the bug is in `ari`?

```
(define coloring (color-graph g locals))
> coloring
(make-immutable-hash (list (cons 'complex-op3051 1) (cons 'complex-if3053 0) (cons 'complex-op30453047 0) (cons 'complex-if30473046 0)))
```


## Hmmm... okay let's modify `instrs` to not use a var for `complex-if3052`
```
(ari instrs coloring)
. . hash-ref: no value found for key
  key: 'complex-if3052
> 
```


## new instrs
```
(define new-instrs
  (list
 (list 'label 'block3058)
 (list 'movq (list 'int 1) (list 'var 'complex-if30473046)) ; notice the change to instrs here
 (list 'movq (list 'var 'complex-if30473046) (list 'var 'complex-op30453047))
 (list 'negq (list 'var 'complex-op30453047))
 (list 'movq (list 'int 46) (list 'var 'complex-op3051))
 (list 'addq (list 'var 'complex-op30453047) (list 'var 'complex-op3051))
 (list 'movq (list 'var 'complex-op3051) (list 'var 'complex-if3053))
 (list 'jmp 'block3054)))
```


## Okay that worked... weeeeeiiiiirrrrrdd 😕 
```
(ari new-instrs coloring)
(list
 (list 'label 'block3058)
 (list 'movq (list 'int 1) (list 'reg 'rbx))
 (list 'movq (list 'reg 'rbx) (list 'reg 'rbx))
 (list 'negq (list 'reg 'rbx))
 (list 'movq (list 'int 46) (list 'reg 'rcx))
 (list 'addq (list 'reg 'rbx) (list 'reg 'rcx))
 (list 'movq (list 'reg 'rcx) (list 'reg 'rbx))
 (list 'jmp 'block3054))
> 
```

**Comment by mfekadu** [comment]
## Let's try again from `make-interference` but with the whole program


**Here is block but you can actually copy paste it into racket interpreter**
```
(define bs `((block
  (,(set 'complex-if3053) ,(set 'complex-if3053) ,(set 'complex-if30483048) ,(set) ,(set))
  (label block3054)
  (movq (var 'complex-if3053) (var 'complex-if30483048))
  (movq (var 'complex-if30483048) (reg rax))
  (jmp conclusion)) 

(block
 (,(set 'complex-if3052)
         ,(set 'complex-if3052)
         ,(set 'complex-if30473046)
         ,(set 'complex-op30453047)
         ,(set 'complex-op30453047)
         ,(set 'complex-op30453047 'complex-op3051)
         ,(set 'complex-op3051)
         ,(set 'complex-if3053)
         ,(set 'complex-if3053))
 (label block3058)
 (movq (var 'complex-if3052) (var 'complex-if30473046))
 (movq (var 'complex-if30473046) (var 'complex-op30453047))
 (negq (var 'complex-op30453047))
 (movq (int 46) (var 'complex-op3051))
(addq (var 'complex-op30453047) (var 'complex-op3051)) ; this one is not cool because they end up both in rbx in the asm output
 (movq (var 'complex-op3051) (var 'complex-if3053))
 (jmp block3054)) 

(block
 (,(set) ,(set) ,(set 'complex-if3052) ,(set 'complex-if3052))
 (label block3060)
 (movq (int 2) (var 'complex-if3052))
 (jmp block3058)) 

(block
 (,(set 'complex-if3050) ,(set 'complex-if3050) ,(set 'complex-if30443044) ,(set 'complex-if30443044 'complex-op3049) ,(set 'complex-op3049) ,(set 'complex-if3053) ,(set 'complex-if3053))
 (label block3055)
 (movq (var 'complex-if3050) (var 'complex-if30443044))
 (movq (int 3) (var 'complex-op3049))
 (addq (var 'complex-if30443044) (var 'complex-op3049))
 (movq (var 'complex-op3049) (var 'complex-if3053))
 (jmp block3054))

(block
 (,(set) ,(set) ,(set 'complex-if3050) ,(set 'complex-if3050))
 (label block3056)
 (movq (int 2) (var 'complex-if3050))
 (jmp block3055))

(block
 (,(set) ,(set) ,(set 'complex-if3050) ,(set 'complex-if3050))
 (label block3057)
 (movq (int 0) (var 'complex-if3050))
 (jmp block3055))

(block
 (,(set) ,(set) ,(set) ,(set) ,(set) ,(set) ,(set) ,(set 'complex-op30423043) ,(set) ,(set) ,(set))
 (label block3061)
 (cmpq (int 1) (int 4))
 (set l (byte-reg al))
 (movzbq (byte-reg al) (var 'complex-op30433042))
 (cmpq (int 0) (var 'complex-op30433042))
 (set e (byte-reg al))
 (movzbq (byte-reg al) (var 'complex-op30423043))
 (cmpq (var 'complex-op30423043) (int 1))
 (jmp-if e block3056)
 (jmp block3057))

(block
 (,(set) ,(set) ,(set 'complex-if3052) ,(set 'complex-if3052))
 (label block3059)
 (movq (int 4) (var 'complex-if3052))
 (jmp block3058))

(block
 (,(set 'x30403040) ,(set 'x30403040) ,(set) ,(set) ,(set 'complex-op30463045) ,(set) ,(set) ,(set))
 (label block3062)
 (cmpq (var x30403040) (int 1))
 (set e (byte-reg al))
 (movzbq (byte-reg al) (var 'complex-op30463045))
 (cmpq (var 'complex-op30463045) (int 1))
 (jmp-if e block3059)
 (jmp block3060))

(block
 (,(set) ,(set) ,(set) ,(set 'x30403040) ,(set 'x30403040) ,(set 'x30403040 'complex-op30413041) ,(set 'x30403040) ,(set 'x30403040) ,(set 'x30403040))
 (label start)
 (movq (int 1) (var x30403040))
 (cmpq (int 0) (var x30403040))
 (set e (byte-reg al))
 (movzbq (byte-reg al) (var 'complex-op30413041))
 (cmpq (var 'complex-op30413041) (int 1))
 (jmp-if e block3061)
 (jmp block3062))))
```


### Oh geez.. how to get locals???
```

```

**Comment by mfekadu** [comment]
## Dang... let's restart the exploration... by running up to `uncover-live`

```bash
> cat compiler.rkt
...
...
(define (compile prog)
          (uncover-live
            (select-instructions
              (uncover-locals
                (explicate-control
                  (rco-prog
                    (uniquify
                      (shrink
                        (typecheck prog)))))))))
...
...
```


```
> ./compiler.rkt programs/if-test2.rc
(program (complex-op1344 complex-op1339 x1338 complex-op1343 complex-if1345 complex-if1342 complex-op1340 complex-if1346 complex-op1341) ((block (#<set: complex-if1346> #<set: complex-if1346> #<set:> #<set:>) (label block1347) (movq (var complex-if1346) (reg rax)) (jmp conclusion)) (block (#<set: complex-if1342> #<set: complex-if1342> #<set: complex-if1346 complex-if1342> #<set: complex-if1346> #<set: complex-if1346>) (label block1348) (movq (int 3) (var complex-if1346)) (addq (var complex-if1342) (var complex-if1346)) (jmp block1347)) (block (#<set:> #<set:> #<set: complex-if1342> #<set: complex-if1342>) (label block1349) (movq (int 2) (var complex-if1342)) (jmp block1348)) (block (#<set:> #<set:> #<set: complex-if1342> #<set: complex-if1342>) (label block1350) (movq (int 0) (var complex-if1342)) (jmp block1348)) (block (#<set: complex-if1345> #<set: complex-if1345> #<set: complex-op1343> #<set: complex-op1343> #<set: complex-if1346 complex-op1343> #<set: complex-if1346> #<set: complex-if1346>) (label block1351) (movq (var complex-if1345) (var complex-op1343)) (negq (var complex-op1343)) (movq (int 46) (var complex-if1346)) (addq (var complex-op1343) (var complex-if1346)) (jmp block1347)) (block (#<set:> #<set:> #<set: complex-if1345> #<set: complex-if1345>) (label block1352) (movq (int 4) (var complex-if1345)) (jmp block1351)) (block (#<set:> #<set:> #<set: complex-if1345> #<set: complex-if1345>) (label block1353) (movq (int 2) (var complex-if1345)) (jmp block1351)) (block (#<set:> #<set:> #<set:> #<set:> #<set:> #<set:> #<set:> #<set: complex-op1340> #<set:> #<set:> #<set:>) (label block1354) (cmpq (int 1) (int 4)) (set l (byte-reg al)) (movzbq (byte-reg al) (var complex-op1341)) (cmpq (int 0) (var complex-op1341)) (set e (byte-reg al)) (movzbq (byte-reg al) (var complex-op1340)) (cmpq (var complex-op1340) (int 1)) (jmp-if e block1349) (jmp block1350)) (block (#<set: x1338> #<set: x1338> #<set:> #<set:> #<set: complex-op1344> #<set:> #<set:> #<set:>) (label block1355) (cmpq (var x1338) (int 1)) (set e (byte-reg al)) (movzbq (byte-reg al) (var complex-op1344)) (cmpq (var complex-op1344) (int 1)) (jmp-if e block1352) (jmp block1353)) (block (#<set:> #<set:> #<set:> #<set: x1338> #<set: x1338> #<set: x1338 complex-op1339> #<set: x1338> #<set: x1338> #<set: x1338>) (label start) (movq (int 1) (var x1338)) (cmpq (int 0) (var x1338)) (set e (byte-reg al)) (movzbq (byte-reg al) (var complex-op1339)) (cmpq (var complex-op1339) (int 1)) (jmp-if e block1354) (jmp block1355))))
```



**MORE READABLE!!!**

```racket
(program
 (complex-op1344 complex-op1339 x1338 complex-op1343 complex-if1345 complex-if1342 complex-op1340 complex-if1346 complex-op1341)
 (
  (block
   (#<set: complex-if1346> #<set: complex-if1346> #<set:> #<set:>)
   (label block1347)
   (movq (var complex-if1346) (reg rax))
   (jmp conclusion))
  (block
   (#<set: complex-if1342> #<set: complex-if1342> #<set: complex-if1346 complex-if1342> #<set: complex-if1346> #<set: complex-if1346>)
   (label block1348)
   (movq (int 3) (var complex-if1346))
   (addq (var complex-if1342) (var complex-if1346))
   (jmp block1347))
  (block
   (#<set:> #<set:> #<set: complex-if1342> #<set: complex-if1342>)
   (label block1349)
   (movq (int 2) (var complex-if1342))
   (jmp block1348))
  (block
   (#<set:> #<set:> #<set: complex-if1342> #<set: complex-if1342>)
   (label block1350)
   (movq (int 0) (var complex-if1342))
   (jmp block1348))
  (block
   (#<set: complex-if1345> #<set: complex-if1345> #<set: complex-op1343> #<set: complex-op1343> #<set: complex-if1346 complex-op1343> #<set: complex-if1346> #<set: complex-if1346>)
   (label block1351)
   (movq (var complex-if1345) (var complex-op1343))
   (negq (var complex-op1343))
   (movq (int 46) (var complex-if1346))
   (addq (var complex-op1343) (var complex-if1346))
   (jmp block1347))
  (block
   (#<set:> #<set:> #<set: complex-if1345> #<set: complex-if1345>)
   (label block1352)
   (movq (int 4) (var complex-if1345))
   (jmp block1351))
  (block
   (#<set:> #<set:> #<set: complex-if1345> #<set: complex-if1345>)
   (label block1353)
   (movq (int 2) (var complex-if1345))
   (jmp block1351))
  (block
   (#<set:> #<set:> #<set:> #<set:> #<set:> #<set:> #<set:> #<set: complex-op1340> #<set:> #<set:> #<set:>)
   (label block1354)
   (cmpq (int 1) (int 4))
   (set l (byte-reg al))
   (movzbq (byte-reg al) (var complex-op1341))
   (cmpq (int 0) (var complex-op1341))
   (set e (byte-reg al))
   (movzbq (byte-reg al) (var complex-op1340))
   (cmpq (var complex-op1340) (int 1))
   (jmp-if e block1349)
   (jmp block1350))
  (block
   (#<set: x1338> #<set: x1338> #<set:> #<set:> #<set: complex-op1344> #<set:> #<set:> #<set:>)
   (label block1355)
   (cmpq (var x1338) (int 1))
   (set e (byte-reg al))
   (movzbq (byte-reg al) (var complex-op1344))
   (cmpq (var complex-op1344) (int 1))
   (jmp-if e block1352)
   (jmp block1353))
  (block
   (#<set:> #<set:> #<set:> #<set: x1338> #<set: x1338> #<set: x1338 complex-op1339> #<set: x1338> #<set: x1338> #<set: x1338>)
   (label start)
   (movq (int 1) (var x1338))
   (cmpq (int 0) (var x1338))
   (set e (byte-reg al))
   (movzbq (byte-reg al) (var complex-op1339))
   (cmpq (var complex-op1339) (int 1))
   (jmp-if e block1354)
   (jmp block1355))))
```

**Comment by mfekadu** [comment]
**Now everything is a `(set 'foo)`**
**make sure all is good by using [diff-checker](https://www.diffchecker.com)**
```racket
`(program
 (complex-op1344 
 complex-op1339 
 x1338 
 complex-op1343 
 complex-if1345 
 complex-if1342 
 complex-op1340
  complex-if1346 
  complex-op1341)
 (
  (block
   (,(set 'complex-if1346) ,(set 'complex-if1346) ,(set) ,(set))
   (label block1347)
   (movq (var complex-if1346) (reg rax))
   (jmp conclusion))
  (block
   (,(set 'complex-if1342) ,(set 'complex-if1342) ,(set 'complex-if1346 'complex-if1342) ,(set 'complex-if1346) ,(set 'complex-if1346))
   (label block1348)
   (movq (int 3) (var complex-if1346))
   (addq (var complex-if1342) (var complex-if1346))
   (jmp block1347))
  (block
   (,(set) ,(set) ,(set 'complex-if1342) ,(set 'complex-if1342))
   (label block1349)
   (movq (int 2) (var complex-if1342))
   (jmp block1348))
  (block
   (,(set) ,(set) ,(set 'complex-if1342) ,(set 'complex-if1342))
   (label block1350)
   (movq (int 0) (var complex-if1342))
   (jmp block1348))
  (block
   (,(set 'complex-if1345) ,(set 'complex-if1345) ,(set 'complex-op1343) ,(set 'complex-op1343) ,(set 'complex-if1346 'complex-op1343) ,(set 'complex-if1346) ,(set 'complex-if1346))
   (label block1351)
   (movq (var complex-if1345) (var complex-op1343))
   (negq (var complex-op1343))
   (movq (int 46) (var complex-if1346))
   (addq (var complex-op1343) (var complex-if1346))
   (jmp block1347))
  (block
   (,(set) ,(set) ,(set 'complex-if1345) ,(set 'complex-if1345))
   (label block1352)
   (movq (int 4) (var complex-if1345))
   (jmp block1351))
  (block
   (,(set) ,(set) ,(set 'complex-if1345) ,(set 'complex-if1345))
   (label block1353)
   (movq (int 2) (var complex-if1345))
   (jmp block1351))
  (block
   (,(set) ,(set) ,(set) ,(set) ,(set) ,(set) ,(set) ,(set 'complex-op1340) ,(set) ,(set) ,(set))
   (label block1354)
   (cmpq (int 1) (int 4))
   (set l (byte-reg al))
   (movzbq (byte-reg al) (var complex-op1341))
   (cmpq (int 0) (var complex-op1341))
   (set e (byte-reg al))
   (movzbq (byte-reg al) (var complex-op1340))
   (cmpq (var complex-op1340) (int 1))
   (jmp-if e block1349)
   (jmp block1350))
  (block
   (,(set 'x1338) ,(set 'x1338) ,(set) ,(set) ,(set 'complex-op1344) ,(set) ,(set) ,(set))
   (label block1355)
   (cmpq (var x1338) (int 1))
   (set e (byte-reg al))
   (movzbq (byte-reg al) (var complex-op1344))
   (cmpq (var complex-op1344) (int 1))
   (jmp-if e block1352)
   (jmp block1353))
  (block
   (,(set) ,(set) ,(set) ,(set 'x1338) ,(set 'x1338) ,(set 'x1338 'complex-op1339) ,(set 'x1338) ,(set 'x1338) ,(set 'x1338))
   (label start)
   (movq (int 1) (var x1338))
   (cmpq (int 0) (var x1338))
   (set e (byte-reg al))
   (movzbq (byte-reg al) (var complex-op1339))
   (cmpq (var complex-op1339) (int 1))
   (jmp-if e block1354)
   (jmp block1355))))
```

**Comment by mfekadu** [comment]
resolved by #41 😄
